### DISCUSSION TITLE
bin/vite binstub does not work as expected (was not generated by bundler error)

### DISCUSSION THREAD

**Discussion by vfonic** [ElMassimo/vite_ruby#75] [Q&A] [answered]
[troubleshooting section]: https://vite-ruby.netlify.app/guide/troubleshooting.html

- [x] I have tried upgrading by running `bundle update vite_ruby`.
- [x] I have read the __[troubleshooting section]__ before opening an issue.

### Description 📖

While installing vite_rails, a new binstub is added to my app's `bin/` directory. When I try to run `bin/vite`, I get the following error (from `bin/vite` script itself):

> Your `bin/bundle` was not generated by Bundler, so this binstub cannot run.
> Replace `bin/bundle` by running `bundle binstubs bundler --force`, then run this command again.

### Reproduction 🐞

I ran `bundle binstubs bundler --force`, which replaced all of my existing binstubs with binstubs that make rails app throw errors on startup (it is no longer possible to start the rails server). Aside from that, this command also made it permanent so that on every `bundle install` all of the binstubs are being regenerated. Also, there are way too many unnecessary binstubs being generated. Almost one binstub for every gem in my `Gemfile`. This is not desirable behavior and should not be a recommended solution.

<details>
<summary>Vite Ruby Info</summary>

_Run `bin/rake vite:info` and provide the output:_

```
bin/vite present?: true
vite_ruby: 1.2.9
vite_rails: 2.0.10
rails: 5.2.4.3
node: v14.16.1
npm: 6.14.12
yarn: 1.22.10
pnpm: 5.17.3
ruby: ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-darwin19]

installed packages:
/Users/viktor/Developer/Ruby/CivicLift
├── vite@2.2.4
└── vite-plugin-ruby@2.0.3
```
</details>

### Logs 📜

_If not providing a reproduction:_

<details>
<summary>Output</summary>

_Run `DEBUG=vite-plugin-ruby:* bin/vite dev` or `DEBUG=vite-plugin-ruby:* bin/vite build` and provide the output:_

```
> DEBUG=vite-plugin-ruby:* bin/vite dev
Your `bin/bundle` was not generated by Bundler, so this binstub cannot run.
Replace `bin/bundle` by running `bundle binstubs bundler --force`, then run this command again.

> DEBUG=vite-plugin-ruby:* bin/vite build
Your `bin/bundle` was not generated by Bundler, so this binstub cannot run.
Replace `bin/bundle` by running `bundle binstubs bundler --force`, then run this command again.

# after changing `bin/bundler` to be compatible with `bin/vite`:
 > DEBUG=vite-plugin-ruby:* bin/vite dev

  vite-plugin-ruby:config {
  vite-plugin-ruby:config   base: '/vite-dev/',
  vite-plugin-ruby:config   build: {
  vite-plugin-ruby:config     emptyOutDir: true,
  vite-plugin-ruby:config     sourcemap: false,
  vite-plugin-ruby:config     assetsDir: 'assets',
  vite-plugin-ruby:config     manifest: true,
  vite-plugin-ruby:config     outDir: '../../../public/vite-dev',
  vite-plugin-ruby:config     rollupOptions: { input: [Object], output: [Object] }
  vite-plugin-ruby:config   },
  vite-plugin-ruby:config   root: '/Users/viktor/Developer/Ruby/CivicLift/app/javascript/entrypoints',
  vite-plugin-ruby:config   server: { host: 'localhost', https: false, port: 3036, strictPort: true },
  vite-plugin-ruby:config   entrypoints: {
  vite-plugin-ruby:config     index: '/Users/viktor/Developer/Ruby/CivicLift/app/javascript/entrypoints/index.tsx'
  vite-plugin-ruby:config   }
  vite-plugin-ruby:config } +0ms


> DEBUG=vite-plugin-ruby:* bin/vite build
Building with Vite ⚡️
vite v2.2.4 building for production...
transforming...
✓ 1340 modules transformed.
2021-05-07T12:32:04.389Z vite-plugin-ruby:config {
  base: '/vite/',
  build: {
    emptyOutDir: true,
    sourcemap: true,
    assetsDir: 'assets',
    manifest: true,
    outDir: '../../../public/vite',
    rollupOptions: { input: [Object], output: [Object] }
  },
  root: '/Users/viktor/Developer/Ruby/CivicLift/app/javascript/entrypoints',
  server: { host: 'localhost', https: false, port: 3036, strictPort: true },
  entrypoints: {
    index: '/Users/viktor/Developer/Ruby/CivicLift/app/javascript/entrypoints/index.tsx'
  }
}
Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db

Why you should do it regularly:
https://github.com/browserslist/browserslist#browsers-data-updating
[vite:css] expected ")".
   ╷
48 │       calc(50% - #{$wallHeaderArrow-mobile-width}}) 0,
   │                                                  ^
   ╵
  app/javascript/scss/_public-wall.scss 48:50  @import
  app/javascript/scss/index.scss 37:9          root stylesheet
file: /Users/viktor/Developer/Ruby/CivicLift/app/javascript/scss/index.scss
Error: expected ")".
   ╷
48 │       calc(50% - #{$wallHeaderArrow-mobile-width}}) 0,
   │                                                  ^
   ╵
  app/javascript/scss/_public-wall.scss 48:50  @import
  app/javascript/scss/index.scss 37:9          root stylesheet
error during build:
Error: expected ")".
   ╷
48 │       calc(50% - #{$wallHeaderArrow-mobile-width}}) 0,
   │                                                  ^
   ╵
  app/javascript/scss/_public-wall.scss 48:50  @import
  app/javascript/scss/index.scss 37:9          root stylesheet
    at Object._newRenderError (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:13172:19)
    at Object._wrapException (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:12998:16)
    at _render_closure1.call$2 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:81115:21)
    at _RootZone.runBinary$3$3 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:27260:18)
    at _FutureListener.handleError$1 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:25816:19)
    at _Future__propagateToListeners_handleError.call$0 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:26114:49)
    at Object._Future__propagateToListeners (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:4536:77)
    at _Future._completeError$2 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:25946:9)
    at _AsyncAwaitCompleter.completeError$2 (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:25600:12)
    at Object._asyncRethrow (/Users/viktor/Developer/Ruby/CivicLift/node_modules/sass/sass.dart.js:4335:17)

Build with Vite failed! ❌
```
</details>

**Comment by ElMassimo** [comment, marked as answer]
[documentation]: https://vite-ruby.netlify.app/guide/troubleshooting.html#troubleshooting

Hi Viktor!

> This is not desirable behavior and should not be a recommended solution.

Although bundler might suggest running `bundle binstubs bundler --force`, that's __not part__ of the recommended installation process—the [documentation] never mentions running `bundle binstubs`.

The `bin/vite` binstub is [generated during the installation process](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/commands.rb#L63-L66) by running bundler.

If there is a problem in the application's bundler setup, you can always run `bundle exec vite` instead.

<hr>

If you would like to get help solving this issue, please use a friendlier tone, and provide additional information, such as your __Bundler version__, the content of `bin/vite`, and the result of `bundle config list`.

Check that `bin/bundle` is available in your project, and that you haven't removed the "This file was generated by bundler" comment from `bin/bundle`, as that's a possible cause for this error.

It could happen that the version of bundler that generated `bin/bundle` is [not compatible](https://github.com/rubygems/bundler/issues/6149#issuecomment-343700345) with the current bundler version you are using (the new one expects a comment, while the old one didn't add it).

**Comment by msakrejda** [comment]
I ran into the same issue--I was also misled by the bundler error message. I found the answer [here](https://stackoverflow.com/questions/11662586/how-to-prevent-bundler-from-generating-binstubs): you can run `bundle config --delete bin` to avoid this behavior.

**Reply by ElMassimo** [reply]
Thanks for sharing!

**Comment by vfonic** [comment]
Hey Massimo!

I'm sorry if my original message was interpreted as in an unfriendly tone. That wasn't my intention.
Once you closed the issue I couldn't reply anymore to the issue and I didn't want to push it by opening another issue as a reply to the original issue.

Let's work together on fixing this issue.
Any info you need, let me know and I'll provide.

I've found a solution since, but now I don't remember what was the solution anymore. :/
I believe it was some bundler configuration that I changed and the binstubs stopped being generated.

Here are the contents of my `~/.bundle/config`:

```sh
~ > cat ~/.bundle/config
---
BUNDLE_BUILD__NOKOGIRI: "--use-system-libraries"
BUNDLE_GEM__COC: "true"
BUNDLE_GEM__MIT: "true"
BUNDLE_GEM__TEST: "rspec"
BUNDLE_GITHUB__HTTPS: "true"
BUNDLE_DEFAULT: "1.17.3"
BUNDLE_FROZEN: "false"
BUNDLE_JOBS: "6"
```

It might be that `BUNDLE_FROZEN` option... 🤔 

I'll try to find out what I did and will let you know.

That being said, maybe there is a better/simpler solution than having to change the bundler config.


EDIT:
Ah, I remembered. I opened the local (in the project root) `.bundle/config` file and removed pretty-much everything.

This is how my `.bundle/config` looks like right now:

```sh
---
BUNDLE_JOBS: "6"
```

**Reply by ElMassimo** [reply]
Sorry, forgot to tag you after moving it to _Discussions_.

From what you and @uhoh-itsmaciek mentioned, I think it's likely that `BUNDLE_BIN` was set when running `bundler binstub bundler --force`.

I'll investigate if that's the case, and see if we can come up with a warning of some sort. At the very least, it makes sense to mention `bundle config --delete bin` in the _Troubleshooting_ section.

The last time I checked, this would only happen in applications where the `bin/bundle` stub was generated with an older version of bundler, it wouldn't happen in new apps.

**Reply by vfonic** [reply]
That could be the case since I used to use Bundler 1.17.3 in this project back then.

**Reply by ElMassimo** [reply]
Running `bundler binstub bundler --force` does not seem to set `BUNDLE_BIN`.

Since this error can happen on any project where `bin/bundle` hasn't been updated in a long time, I added an item in the _Troubleshooting_ guide so that at least the solution is easier to search for.

Thanks!

**Reply by templeman15** [reply]
I just ran into the same issue as the other two in this thread. 

`bundle config --delete bin` fixed it for me! Thanks 😄
